server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -348,6 +361,97 @@ static boolean shouldMinimizeRoundtrips(SearchRequest searchRequest) {
             source.collapse().getInnerHits().isEmpty();
     }
 
+   public static FieldsOptionSourceAdapter createFieldsOptionAdapter(Connection connection, SearchSourceBuilder ccsSearchSource) {
+        Version version = connection.getVersion();
+        if (version.before(Version.V_7_10_0)) {
+            List<FieldAndFormat> fetchFields = ccsSearchSource.fetchFields();
+            if (fetchFields != null) {
+                if (fetchFields.isEmpty() == false) {
+                    String[] includes = fetchFields.stream().map(ff -> ff.field).toArray(i -> new String[i]);
+                    CharacterRunAutomaton unmappedFieldsFetchAutomaton = null;
+                    // We separate the "include_unmapped" field patters with wildcards from the rest in order to use less space in the
+                    // lookup automaton
+                    Map<Boolean, List<String>> partitions = fetchFields.stream()
+                        .map(ff -> ff.field)
+                        .collect(Collectors.partitioningBy((s -> Regex.isSimpleMatchPattern(s))));
+                    List<String> unmappedWildcardPattern = partitions.get(true);
+                    List<String> unmappedConcreteFields = partitions.get(false);
+                    if (unmappedWildcardPattern.isEmpty() == false) {
+                        unmappedFieldsFetchAutomaton = new CharacterRunAutomaton(
+                            Regex.simpleMatchToAutomaton(unmappedWildcardPattern.toArray(new String[unmappedWildcardPattern.size()])),
+                            100000
+                        );
+                    }
+                    final FieldFetcher fieldFetcher = new FieldFetcher(
+                        Collections.emptyMap(),
+                        unmappedFieldsFetchAutomaton,
+                        unmappedConcreteFields
+                    );
+
+                    return new FieldsOptionSourceAdapter() {
+
+                        private boolean removeSourceOnResponse = false;
+
+                        @Override
+                        public void adaptRequest(SearchSourceBuilder source, Consumer<SearchSourceBuilder> sourceConsumer) {
+                            FetchSourceContext fetchSource = source.fetchSource();
+                            // case 1: original request has source: true, but no includes/exclude -> do nothing on request
+                            if (fetchSource != null && fetchSource.fetchSource()) {
+                                if (fetchSource.includes().length == 0 && fetchSource.excludes().length == 0) {
+                                    // do nothing, we can get everything from source and can leave it when translating the response
+                                } else {
+                                    // we probably need to error here
```

Comment:
This question is still open, I think in the case that the original request contains "source" : true and some includes/excluded separate from the "fields" option we cannot simply overwrite them so we should error. That still leaves the emulation option working for most cases including Discovery from Kibana.